pkg/resource/target_reconciler.go

```diff
@@ -0,0 +1,123 @@
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/crossplaneio/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane-runtime/pkg/logging"
+	"github.com/crossplaneio/crossplane-runtime/pkg/meta"
+)
+
+const (
+	targetControllerName   = "kubernetestarget.crossplane.io"
+	targetReconcileTimeout = 1 * time.Minute
+
+	errGetTarget                 = "unable to get Target"
+	errManagedResourceIsNotBound = "managed resource in Target clusterRef is unbound"
+	errUpdateTarget              = "unable to update Target"
+)
+
+// A TargetKind contains the type metadata for a kind of target resource.
+type TargetKind schema.GroupVersionKind
+
+// A TargetReconciler reconciles targets by propagating the secret of the
+// referenced managed resource.
+type TargetReconciler struct {
+	client     client.Client
+	newTarget  func() Target
+	newManaged func() Managed
+
+	propagator ManagedConnectionPropagator
+}
+
+// NewTargetReconciler returns a Reconciler that reconciles KubernetesTargets by
+// propagating the referenced Kubernetes cluster's connection Secret to the
+// namespace of the KubernetesTarget.
+func NewTargetReconciler(m manager.Manager, of TargetKind, with ManagedKind) *TargetReconciler {
+	nt := func() Target { return MustCreateObject(schema.GroupVersionKind(of), m.GetScheme()).(Target) }
+	nr := func() Managed { return MustCreateObject(schema.GroupVersionKind(with), m.GetScheme()).(Managed) }
+
+	// Panic early if we've been asked to reconcile a target or resource kind
+	// that has not been registered with our controller manager's scheme.
+	_, _ = nt(), nr()
+
+	r := &TargetReconciler{
+		client:     m.GetClient(),
+		newTarget:  nt,
+		newManaged: nr,
+		propagator: NewAPIManagedConnectionPropagator(m.GetClient(), m.GetScheme()),
+	}
+
+	return r
+}
+
+// Reconcile a target with a concrete managed resource.
+func (r *TargetReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log.V(logging.Debug).Info("Reconciling", "controller", targetControllerName, "request", req)
+
+	ctx, cancel := context.WithTimeout(context.Background(), targetReconcileTimeout)
+	defer cancel()
+
+	target := r.newTarget()
+	if err := r.client.Get(ctx, req.NamespacedName, target); err != nil {
+		// There's no need to requeue if we no longer exist. Otherwise we'll be
+		// requeued implicitly because we return an error.
+		return reconcile.Result{}, errors.Wrap(IgnoreNotFound(err), errGetTarget)
+	}
+
+	if target.GetWriteConnectionSecretToReference() == nil {
+		// If the ConnectionSecretRef is not set on this Target, we generate a
+		// Secret name that matches the UID of the Target. We implicitly
+		// requeued because of the Target update.
+		target.SetWriteConnectionSecretToReference(&v1alpha1.LocalSecretReference{Name: string(target.GetUID())})
```

Comment:
@negz @muvaf  The `Secret` will be created with a `Name` that matches the `Target` UID. Does that feel like the right pattern to y'all?

Reply:
SGTM!